cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -1992,6 +1988,56 @@ pub(crate) fn define(
         e.enc_32_64_maybe_isap(imul, rec_fa.opcodes(opcodes), *isap);
     }
 
+    // SIMD logical operations
+    for ty in ValueType::all_lane_types().filter(allowed_simd_type) {
+        // and
+        let band = band.bind_vector_from_lane(ty, sse_vector_size);
+        e.enc_32_64(band, rec_fa.opcodes(&PAND));
+
+        // or
+        let bor = bor.bind_vector_from_lane(ty, sse_vector_size);
+        e.enc_32_64(bor, rec_fa.opcodes(&POR));
+
+        // xor
+        let bxor = bxor.bind_vector_from_lane(ty, sse_vector_size);
+        e.enc_32_64(bxor, rec_fa.opcodes(&PXOR));
+    }
+
+    // SIMD bitcast from I32/I64 to the low bits of a vector (e.g. I64x2); this register movement
+    // allows SIMD shifts to be legalized more easily. TODO ideally this would be typed as an
+    // I128x1 but restrictions on the type builder prevent this; the general idea here is that
+    // the upper bits are all zeroed and do not form parts of any separate lane.
+    e.enc_both(
+        bitcast
+            .bind_vector_from_lane(I64, sse_vector_size)
+            .bind(I32),
+        rec_frurm.opcodes(&MOVD_LOAD_XMM),
+    );
+    e.enc64(
+        bitcast
+            .bind_vector_from_lane(I64, sse_vector_size)
+            .bind(I64),
+        rec_frurm.opcodes(&MOVD_LOAD_XMM).rex().w(),
+    );
+
+    // SIMD shift left
+    for (ty, opcodes) in &[(I16, &PSLLW), (I32, &PSLLD), (I64, &PSLLQ)] {
+        let x86_psll = x86_psll.bind_vector_from_lane(ty.clone(), sse_vector_size);
+        e.enc_32_64(x86_psll, rec_fa.opcodes(*opcodes));
+    }
+
+    // SIMD shift right (logical)
+    for (ty, opcodes) in &[(I16, &PSRLW), (I32, &PSRLD), (I64, &PSRLQ)] {
+        let x86_psrl = x86_psrl.bind_vector_from_lane(ty.clone(), sse_vector_size);
+        e.enc_32_64(x86_psrl, rec_fa.opcodes(*opcodes));
+    }
+
+    // SIMD shift right (arithmetic)
+    for (ty, opcodes) in &[(I16, &PSRAW), (I32, &PSRAD)] {
```

Comment:
No 64bit version?

Reply:
Unfortunately not; I'm still trying to figure out how/if to legalize it.